src/main/java/org/broadinstitute/consent/http/service/DatasetRegistrationService.java

```diff
@@ -0,0 +1,426 @@
+import org.broadinstitute.consent.http.models.dataset_registration_v1.ConsentGroup;
+import org.broadinstitute.consent.http.models.dataset_registration_v1.DataLocation;
+import org.broadinstitute.consent.http.models.dataset_registration_v1.DatasetRegistrationSchemaV1;
+import org.broadinstitute.consent.http.models.dataset_registration_v1.NihAnvilUse;
+import org.broadinstitute.consent.http.models.dataset_registration_v1.NihICsSupportingStudy;
+import org.broadinstitute.consent.http.service.dao.DatasetServiceDAO;
+import org.broadinstitute.consent.http.util.gson.GsonUtil;
+import org.glassfish.jersey.media.multipart.FormDataBodyPart;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.BiFunction;
+
+/**
+ * Specialized service class which specifically handles
+ * the process of registering a new dataset into the system,
+ * i.e. creates new datasets.
+ */
+public class DatasetRegistrationService {
+    private final DatasetDAO datasetDAO;
+    private final DacDAO dacDAO;
+    private final DatasetServiceDAO datasetServiceDAO;
+    private final GCSService gcsService;
+
+    public DatasetRegistrationService(DatasetDAO datasetDAO, DacDAO dacDAO, DatasetServiceDAO datasetServiceDAO, GCSService gcsService) {
+        this.datasetDAO = datasetDAO;
+        this.dacDAO = dacDAO;
+        this.datasetServiceDAO = datasetServiceDAO;
+        this.gcsService = gcsService;
+    }
+
+
+    /**
+     * This method takes an instance of a dataset registration schema and creates datasets from it.
+     * There will be one dataset per ConsentGroup in the dataset.
+     *
+     * @param registration The DatasetRegistrationSchemaV1.yaml
+     * @param user The User creating these datasets
+     * @param files Map of files, where the key is the name of the field
+     * @return List of created Datasets from the provided registration schema
+     */
+    public List<Dataset> createDatasetsFromRegistration(
+            DatasetRegistrationSchemaV1 registration,
+            User user,
+            Map<String, FormDataBodyPart> files) throws IOException, SQLException, IllegalArgumentException {
+
+        if (Objects.isNull(dacDAO.findById(registration.getDataAccessCommitteeId()))) {
+            throw new IllegalArgumentException("Invalid DAC id");
+        }
+
+        Map<String, BlobId> uploadedFileCache = new HashMap<>();
+
+        List<DatasetServiceDAO.DatasetInsert> datasetInserts = new ArrayList<>();
+        for (int consentGroupIdx = 0; consentGroupIdx < registration.getConsentGroups().size(); consentGroupIdx++) {
+            datasetInserts.add(createDatasetInsert(registration, user, files, uploadedFileCache, consentGroupIdx));
+        }
```

Comment:
My thought on the lambda was not to replace `createDatasetInsert`, but to call that function inside the lambda instead of using placeholder variables and mutating lists - this is where we often find bugs. The blocker for that was the `IOException`, hence my suggestion to get around that with pre-uploads. One (probably very) minor benefit to the separate upload/dataset insert is that you're not blocking the open db handle while file uploads to GCS are happening. Breaking those two things into discrete chunks will ease the pressure on the handle. 

That said, I can live with this approach :-)

Reply:
> Can you make the DB bits transactional so that the only rollback would be deleting the orphaned files? (and also you could make `createDatasetInsert` no throw a checked exception so that you could call it in a lambda...Man I with there was a way to handle errors in lambdas more cleanly)

The DB bits are transactional, but we have two systems to save to, so combining them is the thing that's problematic to me. We do the following in a single block:
1. Save a dataset with file metadata in postgres (transactional and rollback-able)
2. Save the actual file content in GCS

What I would prefer would be to break them out into two separate rollback-able things so we can fail faster if there are problems in either logical block.

Reply:
To clarify... creating the dataset insert is just creating a record object. So there is no database transaction happening. Basically, what's happening:

1. Upload all files & create dataset insert records at same time (no db changes)
2. Commit all dataset inserts in a rollback-able way at the same time

They are happening sequentially, the only thing I'm missing (if I understand the discourse here) is the ability to rollback files.

Reply:
I added rollback for files. I am gonna add a few tests, and I'll also add some tests for file rollback.